### PR TITLE
Backend (Auth and Database connection)

### PR DESCRIPTION
**Pull request by ArinNigam**
### Description

This pull request introduces significant changes to enhance the application's functionality. The primary updates include integrating MongoDB as the database and implementing user authentication for both signUp and signIn processes.

### Changes Made

1. **Database Integration:**
   - MongoDB has been added as the database backend.
   - Mongoose, a MongoDB ODM, is used for schema modeling and interacting with the database.
   - Necessary configuration settings for connecting to the MongoDB server are included.

2. **Authentication for Signup:**
   - User signup functionality is now secure with the implementation of authentication.
   - Passwords are hashed using bcryptjs before being stored in the database.
   - A new user model schema has been created to store user information securely.

3. **Authentication for Sign-in:**
   - Users can now securely sign in with their credentials.
  

### How to Test

1. Clone this branch locally.
2. Install the required dependencies using `npm install`.
3. Set up a MongoDB server and update the configuration settings in `config.js`.
4. Run the application using `npm start`.
5. Navigate to the signup and sign-in pages and test the functionality using postman.

### Additional Notes

- A new dependency on `mongoose`, `jsonwebtoken` and `bcryptjs` has been added. Ensure these are installed using `npm install` before running the application.
- Existing resources have been updated to work with the MongoDB database.
- Middleware for auth is added which checks whether the token is valid.

### Checklist

- [x] All tests pass.
- [x] Code is properly formatted and follows the project's coding conventions.
- [x] Documentation has been updated to reflect the changes made.
- [x] Security considerations have been taken into account for user authentication.

### CODE REVIEW

**Review by henilp105 — COMMENTED** [review]
Thanks @ArinNigam , I would like to suggest some changes: 
1. add `.DS_Store` to gitignore
2. add temporary varibles/ placeholders instead of real variables
3. don't hardcode mongodb URI , take it as a env var.
4.  consider using yarn or npm managing both for a project becomes difficult/ redundant.

**Comment by ArinNigam** [comment]
1. - [x] add .DS_Store to gitignore
2. - [x] add temporary varibles/ placeholders instead of real variables
3. - [x] don't hardcode mongodb URI , take it as a env var.
4. - [ ] consider using yarn or npm managing both for a project becomes difficult/ redundant.

### Additional Notes
- Apart from this i have added nodemailer which initially was not added in PR.
- I have created authResource for both signUp/signIn using email and password and and an otpResource for logging into the account using an otp send on the mail.
- Will consider using one of yarn or npm after discussing with fellow mentees. For now we are using both.

**Comment by ArinNigam** [comment]
@henilp105 @tusharbansal22 
requesting review for the PR.

File: backend/package.json [hunk 1]
```diff
@@ -5,17 +5,23 @@
   "license": "MIT",
   "description": "Backend server for the smart Insti App",
   "scripts": {
-    "start": "node ./bin/www"
+    "start": "nodemon ./bin/www"
```
Comment (henilp105) [inline backend/package.json:8]: I think for the dev it is good to have nodemon , but for prod it is recomended to have node.

File: backend/resources/authResource.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import User from "../models/user.js";
+import bcryptjs from 'bcryptjs';
+import jwt from 'jsonwebtoken';
+import auth from '../middlewares/auth.js';
+import * as errorMessages from '../constants/errorMessages.js';
+
+const authRouter = express.Router();
+
+// Sign-Up Route
+authRouter.post('/signup',async (req,res)=>{
```
Comment (henilp105) [inline backend/resources/authResource.js:11]: consider adding otp auth during signup.

File: backend/resources/authResource.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        .json({ msg: errorMessages.userNotFound });
+    }
+
+    const isMatch = await bcryptjs.compare(password, user.password);
+    
+    if (!isMatch) {
+      return res.status(400).json({ msg: errorMessages.incorrectPassword });
+    }
+
+    const token = jwt.sign({ id: user._id }, "passwordKey");
```
Comment (henilp105) [inline backend/resources/authResource.js:49]: consider taking this `passwordKey` as an environment variable.

File: backend/resources/otpResource.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  host: process.env.SMTP_HOST,
+  port: process.env.SMTP_PORT,
+  auth: {
+    user: process.env.SMTP_USER,
+    pass: process.env.SMTP_PASS,
+  },
+});
+
+// Store generated OTPs temporarily (for demo purposes)
+const otpStorage = new Map();
```
Comment (henilp105) [inline backend/resources/otpResource.js:33]: consider storing otp's for a user in its mongodb document (variables states are generally lost in serverless architectures.)

File: backend/resources/otpResource.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  // Compare the provided OTP with the stored OTP
+  if (otp === storedOTP) {
+    // Clear OTP from storage after successful verification (for demo purposes)
+    otpStorage.delete(email);
+    return res.json({ message: errorConstants.otpVerfied});
+  }
+
+  res.status(400).json({ error: errorConstants.incorrectOTP });
```
Comment (henilp105) [inline backend/resources/otpResource.js:85]: i think it should be `401`.

File: backend/app.js [hunk 6]
```diff
@@ -3,13 +3,24 @@ import cookieParser from "cookie-parser";
...
+import otpResource from "./resources/otpResource.js";
+import Connection from "./database/db.js";
+import bodyParser from "body-parser";
+
 const app = express();
 
 app.use(logger("dev"));
-app.use(express.json());
+app.use(express.json())
+app.use(bodyParser.json());;
```
Comment (henilp105) [inline backend/app.js:15]: consider using either `express.json` or `bodyParser.json` uing both will be redundant.

**Review by henilp105 — COMMENTED** [review]
Thanks @ArinNigam , I have provided some suggestions. would be happy to merge after fixing them.

**Comment by ArinNigam** [comment]
@henilp105 as suggested by you I have made all the changes in the commit  [Added Changes](https://github.com/OpenLake/Smart-Insti-App/pull/11/commits/67e699c9255fde9a5a6aa77476a5d1a88d92a708)

**Review by henilp105 — APPROVED** [review]

**Pull request merged**